doc/ls_name_value_rd.8
.TH LS_NAME_VALUE_RD "8" "2026-09-29" "lsscsi\-0.33" LSSCSI
.SH NAME
ls_name_value_rd \- get an overview of the contents of a file
.SH SYNOPSIS
.B ls_name_value_rd
[\fI\-\-empty\fR] [\fI\-\-help\fR] \fI\-\-utf8\fR] [\fI\-\-verbose\fR]
[\fI\-\-version\fR] \fIfilename\fR [\fInum_bytes\fR]
.SH DESCRIPTION
.\" Add any additional description here
This utility is a helper for the ls_name_value bash script. This utility
may be useful in its own right as it attempts to give an overview of the
contents of the \fIfilename\fR file. If \fIfilename\fR  is regular,
accessible and contains only 7 bit ASCII character (or perhaps only UTF\-8
encoded sequences) then up to the first \fInum_bytes\fR are output to
stdout. Otherwise some description of \fIfilename\fR or its contents are
output, usually between "<" and ">". For example if \fIfilename\fR is
a "character device" then "<char device>" is output.
.PP
The \fInum_bytes\fR command line argument is optional and is the maximum
number of bytes to read from \fIfilename\fR. Its default value is 80 .
.PP
When errors are detected, sending a description back via stdout is
preferred over more traditional techniques (e.g. sending an error message
to stderr and setting a non\-zero exit code). For example if \fIfilename\fR
is not found then "<not found>" is output to stdout and the exit code is
zero.
.PP
Any null characters (in C: '\\0') found in the contents (value) are skipped
(suppressed) while any ASCII control characters (less than space apart from
the null character) and the 0x7f character are replaced by a space. Note
that all output to stdout is
.B not
ended with a '\\n' character.
.SH OPTIONS
Arguments to long options are mandatory for short options as well.
.TP
\fB\-e\fR, \fB\-\-empty\fR
in many cases the value (contents) associated with \fIfilename\fR is empty.
By default, nothing will output to stdout. When this option is given, if
the value is empty, the string '<empty>' is output to stdout.
.br
Additionally, if this option is given twice, if \fIfilename\fR is regular,
accessible and only contains the '\\n' character (i.e. a linefeed) then "<LF>"
is output to stdout. Note that if this option is not given twice then
that '\\n' control character will be replaced by a space in the output to
stdout.
.TP
\fB\-h\fR, \fB\-\-help\fR
print out the usage message then exit.
.TP
\fB\-u\fR, \fB\-\-utf8\fR
when examining the contents (value) of the \fIfilename\fR file, once the
top bit (MSB) of a byte has been detected, the default value output will be
either "<contains non\-ASCII chars>" or "<might be UTF\-8>". If a non UTF\-8
byte is detected then the whole "value" is considered non\-ASCII. If a
truncated UTF\-8 code point is detected (e.g. truncated due to
\fInum_bytes\fR) then if that is the only (partial) UTF\-8 encoded sequence
then the "value" is considered non\-ASCII, otherwise it is consider UTF\-8.
When this option is given instead of outputting "<might be UTF\-8>" the
UTF\-8 string itself is output to stdout. For the actual UTF\-8 encoded
sequences to be rendered the locale must be set appropriately. For example
on the author's laptop 'env | grep LANG' yields "LANG=en_CA.UTF\-8" and
UTF\-8 encoded sequences are rendered properly.
.TP
\fB\-v\fR, \fB\-\-verbose\fR
increase level or verbosity.
.TP
\fB\-V\fR, \fB\-\-version\fR
prints out the version string then exits.
.SH NOTES
The \fIfilename\fR is opened read\-only and with the O_NONBLOCK flag set.
If the user does have permissions to open \fIfilename\fR
then "<cannot access>" is output to stdout. If the user does have
permission but not read\-only access but does have write\-only access
the "<write only>" is output to stdout.
.PP
Due to that O_NONBLOCK setting when a following read(2) tries to read
from \fIfilename\fR and it is not ready (but might be a bit later) then
that read(2) yields an EAGAIN. This utility will then pause for 10
milliseconds and try again. After 100 retries with the same
outcome (approximately 1 second) this utility will send "<timeout>"
to stdout and exit. Note if any characters were read before that timeout,
and they are 7 bit ASCII, then they may precede that "<timeout>" string.
.PP
This utility "looks through" symbolic links (symlinks). So if \fIfilename\fR
<<<<<<< HEAD
is a symlink to a directory then "<directory>" is output. In the way
=======
is to symlink to a directory then "<directory>" is output. In the way
>>>>>>> 9a2e224 (not github pull #15 in changelog; work on manpages)
this utility works with the ls_name_value bash script, that script takes
care of all the symlink handling. If a symlink is "dangling" then this
utility outputs "<not found>" to stdout.
.PP
As a convenience (e.g. when calling this utility from inside a bash script)
empty command line arguments are ignored. For example in a bash script
calling this utility with a last argument like "${3}" when there is no
third argument to the bash script, then that would invoke this utility with
a "" (empty string) argument. And that would normally be a syntax error.
<<<<<<< HEAD
This utility just ignores such empty arguments. This utility (like most from
this author) uses the GNU getopt_long(3) library function for command
line (invocation) processing. The ls_name_value bash script uses the
getopt(1) command which is similar to getopt_long().
=======
This utility just ignores such empty arguments. The utility (like most from
this author) uses the GNU getopt_long() library function, or from inside a
bash script uses the getopt(1) command.
>>>>>>> 9a2e224 (not github pull #15 in changelog; work on manpages)
.SH AUTHORS
Written by D. Gilbert
.SH COPYRIGHT
Copyright \(co 2026 Douglas Gilbert
.br
This software is distributed under a BSD\-2\-Clause license. There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.SH "SEE ALSO"
.B ls_name_value(this package)


ChangeLog
Linux utility for listing SCSI and NVMe devices.

Changelog for pre-release lsscsi-0.33.5 [20260929] [svn: r232]
  - lsscsi: fix off-by-one in --sysfsroot sg_strscpy
    Paul Evans github PR #19
  - ls_name_value: add --absolute option to show symlink's
    target with an absolute path on following line
    - small shellcheck cleanup
  - ls_name_value_rd: small manpage cleanup
  - lsscsi: '-V' (show version) still said "pre-release"
    in the 0.33 release (r228); remove "pre-"

Changelog for release lsscsi-0.33 [20260913] [svn: r228]
  - bounds checking fixes, see lsscsi pull #15 [github]
  - Fix multipath WWID inconsistencies in lsscsi#11 [github]
  - fix Transport type FC and FCOE output inconsistencies
    [github #9]
  - Change default output trailing space after each device
    entry [github #8]
  - fixes for SBP (Firewire) host in transport_h_init()
    [github #7]
  - fix path construction for SAS end device [github #5]
  - add --json[=JO] and --js-file=JFN for JSON output
    - changed short form syntax to -j[=JO]
  - add cmake support and keep autotools
    - can build deb and rpm packages with 'cpack .'
    - align build config with autotools [github pull #14]
    - detect compiler and use for gcc/clang warnings
    - clean up C/C++ issue highighted by those warnings
  - ls_name_value: new bash script with manpage to help
    with sysfs overview,
    - change the directory indication from a value of
      "-->" to its name surrounded by square brackets
    - add new helper command: ls_name_value_rd and manpage
    - add '-f' option for native directory order listing
    - use '--' option and other measures for filenames
      starting with '-'
    - fix portability issue and misc fd fix [github pull #13]
  - add --sysroot=AR_PT option for compatibilty with lsblk,
    lsmem and lscpu
  - start to work on 'missing NVME device nodes' issue
    [https://github.com/doug-gilbert/lsscsi/issues/4]
  - change '-w' output to step over designator type prefix
    (e.g. '3' for NAA)
  - add '-ww' to mean: really use
    /dev/disk/by-id/wwn-* rather than .../by-id/scsi-*
  - document that --scsi_id (-i) now is prefixed by a
    "from whence" character from "328S10"
  - add '-ii' to mean show identifier without that prefix
  - clean up '-t' output for SAS (RAID) logical disks
  - improve handling of smartpqi scsi host
  - limit recursion depth of get_disk_scsi_id()
  - "nvme-generic" devices (e.g. /dev/ng0n1) now processed
    and shown when the --generic option is given
  - with NVMe device single line output, reduce model
    string length to 30 (was 41) to align columns better
  - add "pseudo_0" transport for the scsi_debug driver
    which is named after /sys/devices/pseudo_0
  - build_debian.sh: build without signatures
    - rename to debian_build.sh to minimize files starting
      with "build". Suggest "build" is for out-of-tree builds
  - debian: migrate from cdbs to debhelper [github pull #12]
  - add JSON builder code borrowed from sg3_utils
  - import sg_json.[hc], sg_pr2serr.[hc] and sg_unaligned.h
    from sg3_utils svn revision 1142
  - replace strcpy() with sg_strscpy()
  - remove the imtermediate files generated by ./autogen.sh
    - this reduces size of svn and git repositories but still
      plan to have these intermediate files in release tarballs

Changelog for released lsscsi-0.32 [20210505] [svn: r167]
  - improve NVMe device parsing (e.g. /dev/nvme0c1n2)
  - print nr_hw_queues when available for SCSI hosts
  - Remove blank line after NVMe device name with -HL
  - collect_disk_wwn_nodes: Fix WWN string copy
  - make WWN printing for NVMe more consistent with
    output from SCSI devices (e.g. with -u and -t)
  - logic to select best SCSI id (--scsi_id) to output
  - clean up warnings for gcc-10
  - build with autoconf 2.70

Changelog for released lsscsi-0.31 [20200220] [svn: r160]
  - cleanup gcc-8 warnings, no cleanup needed for gcc-9
  - fix issue where host managed ZBC devices don't
    output their size. [Fix also for RBC and CD/DVD.]
  - exclude NVMe listings when --classic given
  - fix hex counting issues
  - supply "-" for generic NVMe device one line output
    so 'lsscsi -gb' output is consistent

Changelog for released lsscsi-0.30 [20180612] [svn: r154]
  - add support for NVMe devices and controllers
    - to build without: ./configure --disable-nvme-supp
    - deselect at runtime: lsscsi --no-nvme
    - deselect SCSI devices at runtime: lsscsi N
  - add --brief for tuple + device_name(s) only
  - add --pdt (-D) for device type in hex
  - extend --size (-s) so when given three times
    the size as a logical block count is output
  - add --sz-lbs (-S) that is equivalent to '-sss'
    when used twice adds comma then logical block size
  - '-w' now decodes 128 bit WWN without truncation
  - /dev/disk/by-id/wwn- is not guaranteed to be
    persistent (or stable); instead use
    /dev/disk/by-id/scsi-
  - '-t' on a FC host was not printing the comma
    separator resulting in garbled output, fix
  - debian: bump compat file contents from 7 to 10

Changelog for released lsscsi-0.29 [20160514] [svn: r137]
  - '-u' now decodes locally assigned UUIDs (spc5r08)
  - as last try use T10 Vendor ID for lu name
  - if no lu name found, print 'none'
  - change '-uuu' to output the full lu name followed
    by the normal fields (which were skipped before)
  - add 'U' option, same action as '-uuu'
  - '-UU' prefixes lu names with 'eui.', 'naa.', etc
  - if '-s' given twice, lu size is base 2 related
  - if very long [h:c:t:l] then append space
  - print_enclosure_device() for FCP may be useless,
    comment out while checking ...
  - with '-t' print 0x0000000000000000 for non-SAS
    device in SAS domain
  - autogen.sh: upgrade to 20091223 version
  - automake: upgrade to 1.15 (ubuntu 16.04)

Changelog for released lsscsi-0.28 [20140930] [svn: r120]
  - fix handling of scsi_level 0 (no compliance)
  - add SRP transport identifier

src/ls_name_value_rd.c
/*
 * Copyright (c) 2026 Douglas Gilbert.
 * All rights reserved.
 *
 * SPDX-License-Identifier: BSD-2-Clause
 */

/*
 * This is a helper C function for the ls_name_value script that has its
 * own manpage. Inside a script it is difficult to try and read the contents
 * of an arbitrary filename. For example it may not be a regular file, it
 * report as a regular file but hangs (e.g. /proc/kmsg), it is regular but
 * contains binary or the (null) character, plus several other corner cases.
 * The Linux timeout command takes 100 milliseconds even when the contents is
 * provided immediately, making the use of timeout unsuitable.
 * So this command does the "hard work" in C. The idea is not to give an
 * exact rendering of the contents of a file, but give a useful single line
 * overview. Null characters ('\0') are squashed, ASCII control characters
 * and 0x7f are replaced by ' ' while any regular file contents (read up to
 * <num_bytes> bytes) with no top bits set are output to stdout. If the top
 * bit is set in any byte then the output is: <contains non-ASCII chars> .
 * As a bonus, if the top bit is set a check is made for a UTF-8 sequence
 * and if only valid UTF-8 codepoints are found <might be UTF-8> is output.
 * UTF-8 has a corner case where a codepoint is truncated by <num_bytes>:
 * if that happens on the first UTF-8 sequence then the output is
 * <contains non-ASCII chars>.
 * Output goes to stdout with no trailing \n, only errors associated with
 * the command line syntax cause error messages to be sent to stderr.
 *
 * The command line is shown in the usage() and show_help() functions below.
 */

#define _XOPEN_SOURCE 600
#ifndef _GNU_SOURCE
#define _GNU_SOURCE 1
#endif

#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <string.h>
#include <fcntl.h>
#include <errno.h>
#include <stdbool.h>
#include <time.h>
#include <sys/stat.h>
#include <getopt.h>     /* non-standard, trips up AIX */

static const char * version_str = 
                     "0.33.5 (pre 0.34)  20260928 [svn: r231]";

static const char * my_name = "ls_name_value_rd";

static const char * op_not_permit = "operation not permiited";
static const char * permiss_denied = "permission denied";

static int num_retries = 100;    /* 10 milliseconds per try */

#define DEF_BUFF_SZ 256
#define DEF_CONTENT_NUM_BYTES 80


static struct option long_options[] = {
    {"empty", no_argument, 0, 'e'},
    {"help", no_argument, 0, 'h'},
    {"utf8", no_argument, 0, 'u'},
    {"verbose", no_argument, 0, 'v'},
    {"version", no_argument, 0, 'V'},
};

static void
usage()
{
    fprintf(stderr, "Usage: %s [-e] [-h] [-u] [-v] [-V] <filename> "
            "[<num_bytes>]\n", my_name);
}

static void
show_help()
{
    usage();
    fprintf(stderr, "\n    -e | --empty    output <empty> if contents "
            "are empty; use twice to\n");
    fprintf(stderr, "                    additionally output <LF> "
            "instead of contents='\\n'\n");
    fprintf(stderr, "    -h | --help     show this help message then "
            "exit\n");
    fprintf(stderr, "    -u | --utf8     if contents passes utf8 test, "
            "output it\n");
    fprintf(stderr, "    -v | --verbose  increase debug output\n");
    fprintf(stderr, "    -V | --version  show version string then exit\n");
    fprintf(stderr, "    <filename>      name of file to read\n");
    fprintf(stderr, "    <num_bytes>     maximum number of bytes to ");
    fprintf(stderr, "read [def: %d]\n\n", DEF_CONTENT_NUM_BYTES);
    fprintf(stderr, "Read <num_bytes> from <filename> skipping any null ");
    fprintf(stderr, "characters. Any\nbytes with the top bit set will ");
    fprintf(stderr, "cause <contains non-ASCII chars>\nto be output to ");
    fprintf(stderr, "stdout. Empty files (and those with null chars\nup ");
    fprintf(stderr, "to their EOF) output <empty>. Output to stdout does ");
    fprintf(stderr, "not\nhave a trailing \\n .\n");
}

/* Check if a sequence of bytes is valid UTF-8; if so return true and if
 * skip_nump is NON-NULL then write the number of bytes it stepped over. A
 * sequence of bytes for a UTF-8 codepoint may be truncated by sz . Return
 * true in that case if it was valid to that point. Returns false
 * if sequence is invalid, if skip_nump is NON-NULL write the number of
 * bytes stepped over at the point of failure. If skip_nump is NON-NULL
 * then *skip_nump is always >= 1 even if sz<=0 or s[0]=='\0' in which cases
 * true is returned. */
bool
utf8_valid_seq(const char *s, size_t sz, int * skip_nump)
{
    if ((sz > 0) && s[0]) {
        unsigned char c = (unsigned char)s[0];
        size_t len, j;

        if (c < 0x80)
            len = 1;
        else if ((c >> 5) == 0x6)
            len = 2;
        else if ((c >> 4) == 0xE)
            len = 3;
        else if ((c >> 3) == 0x1E)
            len = 4;
        else {
            if (skip_nump)
                *skip_nump = 1;
            return false;
        }

        for (j = 1; (j < len) && (j < sz); j++)
        {
            if ((s[j] & 0xC0) != 0x80) {
                if (skip_nump)
                    *skip_nump = j;
                return false;
            }
        }
        if (skip_nump)
            *skip_nump = j;
    } else if (skip_nump)
        *skip_nump = 1;
    return true;
}

int
main(int argc, char * argv[])
{
    int res, c, fd, off, len, b1len, k, j;
    int empty = 0;
    int num_bytes = DEF_CONTENT_NUM_BYTES;
    int utf8 = 0;
    int verbose = 0;
    int err = 0;
    int skip = 0;
    int num_utf8_codepoi = 0;
    int read_einval = 0;
    bool reg_file = false;
    bool non_ascii = false;
    bool incomplete_utf8 = false;
    bool not_retry = false;
    const char * fn = NULL;
    const char * num_bytes_str = NULL;
    char b1[DEF_BUFF_SZ + 2];
    char b2[DEF_BUFF_SZ + 2];
    char * b1p = b1;
    char * b2p = b2;
    struct stat st;
    /* setup duration for 10 milliseconds per EAGAIN from read() */
    struct timespec duration = { 0, 10 * 1000 * 1000 /* ns */};

    while (1) {
        c = getopt_long(argc, argv, "ehuvV", long_options, NULL);
        if (c == -1)
            break;

        switch (c) {
        case 'e':
            ++empty;
            break;
        case 'h':
            show_help();
            return 0;
        case 'u':
            ++utf8;
            break;
        case 'v':
            ++verbose;
            break;
        case 'V':
            fprintf(stderr, "%s version: %s\n", my_name, version_str);
            return 0;
        default:
            fprintf(stderr, "unrecognised option code: 0x%02x\n", c);
            usage();
            return EINVAL;

        }
    }
    while (optind < argc) {
        const char * avp = argv[optind];
        bool ok = true;

        if (NULL == fn) {
            fn = avp;
            optind++;
        } else if (NULL == num_bytes_str) {
            num_bytes_str = avp;
            optind++;
        } else if (optind < argc) {
            do {
                if (strlen(avp) > 0) {
                    ok = false;
                    fprintf(stderr, "Unexpected extra argument: %s\n", avp);
                } else if (verbose > 3 )
                    fprintf(stderr, "Empty command line argument at "
                            "optind=%d\n", optind);
                if (++optind >= argc)
                    break;
                avp = argv[optind];
            } while (true);
            if (! ok) {
                fprintf(stderr, "\n");
                usage();
                return EINVAL;
            }
        }
    }

    if (num_bytes_str) {
        if (1 != sscanf(num_bytes_str, "%d", &num_bytes)) {
            fprintf(stderr, "%s: bad <num_bytes>\n\n", my_name);
            usage();
            return EINVAL;
        }
        if (num_bytes < 0) {
            fprintf(stderr, "%s: negative <num_bytes>\n\n", my_name);
            usage();
            return EINVAL;
        }
    }
    if (NULL == fn) {
        fprintf(stderr, "%s: filename missing\n\n", my_name);
        usage();
        return EXDEV;  /* arbitrary error code for internal error */
    }
    if (num_bytes > DEF_BUFF_SZ) {
        b1p = (char *)malloc(num_bytes + 2);
        b2p = (char *)malloc(num_bytes + 2);
        if ((NULL == b1p) || (NULL == b2p)) {
            fprintf(stderr, "%s: out of memory\n", my_name);
            if (b1p)
                free(b1p);
            if (b2p)
                free(b2p);
            return ENOMEM;
        }
    }
    if (verbose > 0)
        fprintf(stderr, "filename: %s, num_bytes=%d, empty=%d, "
                "verbose=%d\n", fn, num_bytes, empty, verbose);

    /* Note; O_NONBLOCK that leads to EAGAIN errors on the later read()s */
    fd = open(fn, O_RDONLY | O_NONBLOCK);
    if (fd < 0) {
        err = errno;
        if (verbose > 2)
            fprintf(stderr, "open(): errno=%d\n", err);
        if (ENOENT == err)
            printf("<not found>");
        else if (EACCES == err) {
            if (0 == access(fn, W_OK))
                printf("<write only>");
            else
                printf("<cannot access>");
        } else if (EPERM == err) {
            if (verbose > 0)
                fprintf(stderr, "<open: %s>\n", op_not_permit);
            printf("<%s>", op_not_permit);
        } else if (EINTR == err) {
            if (verbose > 0)
                fprintf(stderr, "<open: interrupted>\n");
            printf("<interrupted>");
        } else if (EIO == err) {
            if (verbose > 0)
                fprintf(stderr, "<open: IO error>\n");
            printf("<IO error>");
        } else if (ELOOP == err)
            printf("<too many symlinks>");
        else if (EINVAL == err) {
            if (verbose > 0)
                fprintf(stderr, "<open: EINVAL>\n");
            printf("<invalid argument>");
            /* printf("<bad character in filename>"); */
        } else
            printf("<open() errno=%d>", err);
        err = 0;
        goto cleanup;
    }
    if (fstat(fd, &st) < 0) {
        err = errno;

        if (verbose > 2)
            fprintf(stderr, "fstat(): errno=%d\n", err);
        if (EBADF == err) {
            if (verbose > 0)
                fprintf(stderr, "<fstat() bad file descriptor>\n");
            err = 0;
            goto cleanup;
        } else {
            if (verbose > 0)
                fprintf(stderr, "<fstat() errno=%d>\n", err);
            printf("<fstat() errno=%d>", err);
            err = 0;
            goto cleanup;
        }
    }
    switch (st.st_mode & S_IFMT) {
    case S_IFREG:
        reg_file = true;
        break;
    case S_IFBLK:
        printf("<block device>");
        break;
    case S_IFCHR:
        printf("<char device>");
        break;
    case S_IFDIR:
        printf("<directory>");
        break;
    case S_IFIFO:
        printf("<named pipe>");         /* FIFO/pipe */
        break;
    case S_IFLNK:
        printf("<symlink>");            /* shouldn't happen */
        break;
    case S_IFSOCK:
        printf("<socket>");
        break;
    default:
        printf("<fstat: unknown type>");
        break;
    }
    if (! reg_file)
        goto cleanup;

    for (len = 0, off = 0, k = 0; k < num_retries; not_retry ? k : ++k) {
        err = 0;
        not_retry = false;
        /* this read() should be fast due to O_NONBLOCK */
        res = read(fd, b1p + off, num_bytes - off);
        if (res > 0) {
            len += res;
            if (len >= num_bytes)
                break;
            off += res;
            not_retry = true;
        } else if (res == 0)
            break;
        else {
            err = errno;
            if (EAGAIN == err) {
                if (verbose > 3)
                    fprintf(stderr, "read: EAGAIN\n");
                nanosleep(&duration, NULL);     /* 10 ms */
                err = 0;
                continue;
            }
            if (verbose > 2)
                fprintf(stderr, "read(): errno=%d\n", err);
            if (EBADF == err)
                printf("<read() bad file descriptor>");
            else if (EINTR == err) {
                if (verbose > 0)
                    fprintf(stderr, "<read: interrupted system call>\n");
                printf("<read interrupted>");
            } else if (EIO == err) {
                if (verbose > 0)
                    fprintf(stderr, "<read: IO error>\n");
                printf("<IO error>");
            } else if (EPERM == err) {
                if (verbose > 0)
                    fprintf(stderr, "<read: %s>\n", op_not_permit);
                printf("<%s>", op_not_permit);
            } else if (EACCES == err) {
                if (verbose > 0)
                    fprintf(stderr, "<read: %s>\n", permiss_denied);
                printf("<%s>", permiss_denied);
            } else if (EINVAL == err) { /* sysfs special, treat like empty */
                ++read_einval;
                break;
            } else
                printf("<read() errno=%d>", err);
            err = 0;
            goto cleanup;
        }
    }           /* end of for loop over retries */
    b1len = len;
    if (k >= num_retries) {
        err = 0;
        if (len > 0)
            printf("%.*s <timeout>", len, b1p);
        else
            printf("<timeout>");
        goto cleanup;
    }
    for (k = 0, j = 0; k < len; ++k) {
        unsigned char cc = *(b1p + k);

        incomplete_utf8 = false;
        if ('\0' == cc)
            continue;   /* squash the null character */
        if (cc < ' ')
            *(b2p + j++) = ' ';/* replace ASCII control characters with ' ' */
        else if (cc > 0x7e) {
            if (0x7f == cc)
                *(b2p + j++) = ' ';
            else {      /* top bit set */
                if (utf8_valid_seq(b1p + k, len - k, &skip)) {
                    if (skip >= (len - k))
                        incomplete_utf8 = true;
                    k += (skip - 1);
                    ++num_utf8_codepoi;
                } else {
                    non_ascii = true;
                    break;
                }
            }
        } else
            *(b2p + j++) = cc;
    }                   /* end of for loop over input characters */
    if (non_ascii) {
        printf("<contains non-ASCII chars>");
        goto cleanup;
    } else if (num_utf8_codepoi > 0) {
        if ((1 == num_utf8_codepoi) && incomplete_utf8)
            printf("<contains non-ASCII chars>");       /* marginal ... */
        else {
            if (utf8 > 0)
                printf("%.*s", len, b1p);
            else
                printf("<might be UTF-8>");
        }
        goto cleanup;
    }
    len = j;
    if (0 == len) {
        if (empty > 0)
            printf("<empty>");
        else {
            ;   /* printf("");   gives warning */
        }
    } else if ((empty > 1) && (1 == b1len) && (b1p[0] == '\n')) {
        printf("<LF>");
    } else
        printf("%.*s", len, b2p);
    err = 0;

cleanup:
    if (fd >= 0)
        close(fd);
    if (b2p != b2)
        free(b2p);
    if (b1p != b1)
        free(b1p);
    if ((verbose > 0) && (read_einval > 0))
        fprintf(stderr, "%s: read_einval=%d\n", my_name, read_einval);
    return err;
}
